### PR TITLE
[Keyboard] Add kudox keyboard

### PR DESCRIPTION
**Pull request by x1-**
## Description

Adds support for Kudox Keyboard by [Kumao Kobo](https://github.com/kumaokobo).

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Adds support for Kudox Keyboard

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/kudox/rules.mk [hunk 1]
```diff
@@ -0,0 +1,73 @@
+# MCU name
+#MCU = at90usb1286
```
Comment (fauxpark) [inline keyboards/kudox/rules.mk]:
```suggestion
```

File: keyboards/kudox/rules.mk [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
+
+
+# Boot Section Size in *bytes*
+#   Teensy halfKay   512
+#   Teensy++ halfKay 1024
+#   Atmel DFU loader 4096
+#   LUFA bootloader  4096
+#   USBaspLoader     2048
+OPT_DEFS += -DBOOTLOADER_SIZE=4096
```
Comment (fauxpark) [inline keyboards/kudox/rules.mk]: If you know the bootloader for this keyboard, please replace this block with:

```make
# Bootloader selection
#   Teensy       halfkay
#   Pro Micro    caterina
#   Atmel DFU    atmel-dfu
#   LUFA DFU     lufa-dfu
#   QMK DFU      qmk-dfu
#   atmega32a    bootloadHID
BOOTLOADER = atmel-dfu
```

File: keyboards/kudox/rev1/rev1.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include "../kudox.h"
+
+#include "quantum.h"
+
+
+#ifdef USE_I2C
```
Comment (fauxpark) [inline keyboards/kudox/rev1/rev1.h]: Pretty sure this whole block is unnecessary, as split common should be setting all of this up (I think).

File: keyboards/kudox/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,81 @@
+#include QMK_KEYBOARD_H
+
+extern keymap_config_t keymap_config;
```
Comment (fauxpark) [inline keyboards/kudox/keymaps/default/keymap.c]:
```suggestion
```

This isn't used anywhere.

**Comment by x1-** [comment]
@fauxpark 
Thank you for reviewing my codes!
I modified the codes you pointed out. (I am glad to see.)

File: keyboards/kudox/rev1/config.h [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+// wiring of each half
+#define MATRIX_ROW_PINS { D4, D7, E6, B4, B5 }
+#define MATRIX_COL_PINS { F5, F6, F7, B1, B3, B2, B6 }
+// #define MATRIX_COL_PINS { B6, B2, B3, B1, F7, F6, F5} //uncomment this line and comment line above if you need to reverse left-to-right key order
+
+/* COL2ROW or ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+#define CATERINA_BOOTLOADER
```
Comment (fauxpark) [inline keyboards/kudox/rev1/config.h]:
```suggestion
```

This define doesn't do anything; there is no code that checks for it.

File: keyboards/kudox/rev1/rules.mk [hunk 6]
```diff
@@ -0,0 +1 @@
+BACKLIGHT_ENABLE = no
```
Comment (fauxpark) [inline keyboards/kudox/rev1/rules.mk]:
```suggestion
```

`BACKLIGHT_ENABLE` is already set to `no` in the keyboard-level rules.mk, so this isn't necessary.

File: keyboards/kudox/keymaps/default/rules.mk [hunk 7]
```diff
@@ -0,0 +1,2 @@
+RGBLIGHT_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/kudox/keymaps/default/rules.mk]:
```suggestion
```

And here too.

File: keyboards/kudox/keymaps/default/config.h [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+/* Use I2C or Serial, not both */
+#define USE_SERIAL
```
Comment (fauxpark) [inline keyboards/kudox/keymaps/default/config.h]: I think all of this stuff should actually go in the keyboard config.h.

**Comment by x1-** [comment]
@fauxpark 
Thank you for your fastest revewing!
I modified the codes you pointed out.

**Comment by mechmerlin** [comment]
Your default keymap looks fairly complex for a default. Is this the intended keymap or is this your keymap?

File: keyboards/kudox/rev1/rev1.h [hunk 9]
```diff
@@ -0,0 +1,25 @@
+#pragma once
+
+#include "../kudox.h"
```
Comment (noroadsleft) [inline keyboards/kudox/rev1/rev1.h]: The keyboard level is added as part of the PATH, so you don't need to specify the directory one level up here.

```suggestion
#include "kudox.h"
```

File: keyboards/kudox/info.json [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+                {"label":"N", "x":10.75, "y":3},
+                {"label":"M", "x":11.75, "y":3},
+                {"label":",", "x":12.75, "y":3},
+                {"label":".", "x":13.75, "y":3},
+                {"label":"Shift", "x":14.75, "y":3, "w":2.25},
+                {"label":"Alt", "x":0, "y":4, "w":1.25},
+                {"label":"'", "x":1.25, "y":4},
+                {"label":"-", "x":2.25, "y":4},
+                {"label":"=", "x":3.25, "y":4},
+                {"label":"\u2190", "x":12.75, "y":4},
```
Comment (noroadsleft) [inline keyboards/kudox/info.json]: Your QMK Configurator implementation breaks down here, and will lead to users' keys being assigned out of place on the bottom row.

Lines 64-73 should be replaced with:

```
                {"label":"layer", "x":4.25, "y":4.33, "w":1.25},
                {"label":"GUI", "x":5.5, "y":4.67},
                {"label":"Space", "x":6.5, "y":5},
                {"label":"Enter", "x":9.5, "y":5},
                {"label":"Del", "x":10.5, "y":4.67},
                {"label":"layer", "x":11.5, "y":4.33, "w":1.25},
                {"label":"\u2190", "x":12.75, "y":4},
                {"label":"\u2193", "x":13.75, "y":4},
                {"label":"\u2191", "x":14.75, "y":4},
                {"label":"\u2192", "x":15.75, "y":4, "w":1.25}
```

**Comment by x1-** [comment]
@mechmerlin 
 
> Your default keymap looks fairly complex for a default. Is this the intended keymap or is this your keymap?

Well... I would like to modify more simple.

**Comment by x1-** [comment]
@noroadsleft 
Thank you for reviewing my code!
I modified the codes you pointed out.
https://github.com/qmk/qmk_firmware/pull/6459/commits/eab2f0106297647d75c69836cec7048a9361cb87
https://github.com/qmk/qmk_firmware/pull/6459/commits/6a5f82542b8127d716c8ec457a079d6069da3f9a

**Comment by x1-** [comment]
@drashna 
Thank you for reviewing my codes!
I separated the default keymap and the keymap for myself.
https://github.com/qmk/qmk_firmware/pull/6459/commits/27cd93bebb92ec96f1df0cebe23a8f0a8caf3652

**Comment by x1-** [comment]
@mechmerlin 
Thank you for reviewing my codes!
I separated the default keymap and the keymap for myself.
https://github.com/qmk/qmk_firmware/pull/6459/commits/27cd93bebb92ec96f1df0cebe23a8f0a8caf3652

**Comment by x1-** [comment]
@drashna Thank you for reviewing!
I am very glad if you see this!

**Pull request merged**